EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
+  
+def add_read_access_events(address: bytes20, sub_key: bytes31, leaf_key: byte):
+  add_access_event(address, sub_key, leaf_key, False)
 ```
-(address, 0, BASIC_DATA_LEAF_KEY)
-```
-
-Note: a non-value-bearing `SELFDESTRUCT`, `*CALL` or `CALLCODE`, targeting a precompile or a system contract, will not cause the `BASIC_DATA_LEAF_KEY` to be added to the witness.
-
-If a `*CALL`, `CALLCODE` or `SELFDESTRUCT` is value-bearing (ie. it transfers nonzero wei), whether or not the `callee` is a precompile or a system contract, process this additional access event:
 
-```
-(caller, 0, BASIC_DATA_LEAF_KEY)
-```
+The consequence of this model, is that it created group-locality in the state, with each "group" containing 256 contiguous leaves.
 
-Note: when checking for the existence of the `callee`, the existence check is done by validating that there is an extension-and-suffix tree at the corresponding stem, and does not rely on `CODEHASH_LEAF_KEY`.
+An event will be triggered:
 
-When calling `EXTCODEHASH`, process the access event:
+ * Every time a group is accessed
+ * Every time a single leaf is accessed
 
-```
-(address, 0, CODEHASH_LEAF_KEY)
-```
-
-Note that precompiles and system contracts are excluded, as their hashes are known to the client.
-
-When a contract is created, process these access events:
-
-```
-(contract_address, 0, BASIC_DATA_LEAF_KEY)
-(contract_address, 0, CODEHASH_LEAF_KEY)
-```
-
-#### Access events for storage
-
-`SLOAD` and `SSTORE` opcodes with a given address and key process an access event of the form
-
-```
-(address, tree_key, sub_key)
-```
+The model makes a distinction between read and write events. As we will see later, gas costs are associated to these events.
 
-Where `tree_key` and `sub_key` are computed as `tree_key, sub_key = get_storage_slot_tree_keys(address, key)`
+#### Atomicity
 
-#### Access events for code
+The calls to `charge_gas()` are here to delimit the atomic boundaries of an operation. All the events emitted between two consecutive calls to `charge_gas` are charged together. If there isn't enough gas to pay for all events, none of the associated values are added to the witness.
 
-In the conditions below, “chunk chunk_id is accessed” is understood to mean an access event of the form
+#### Access events for account headers
 
-```
-(address, (chunk_id + 128) // 256, (chunk_id + 128) % 256)
+```python3
+# Emit all access events given an instruction. It is meant to be called before
+# the gas charging method for a given instruction is called, as these events
+# are subsequently used to process witness-related gas costs.
+def emit_access_events(opcode: int, stack: Stack, contract: Contract, new_contract: Optional[Contract]):
+  match opcode:
+    case OP_CALL | OP_DELEGATECALL | OP_STATICCALL | OP_CALLCODE | OP_SELFDESTRUCT:
+      address = stack.peek(1)
+
+      # XXX the previous version of the spec says OP_CALLCODE is concerned, but the geth
+      # code says it isn't.
```

Comment:
Inconsistent variable name (sometimes `opcode` sometimes `op` for the same variable)